Add IsReadableStreamDisturbed predicate

tyoshino opened the pull request:
See the suggestion in the OP of #378.

tyoshino commented:
The disturbed flag is never changed when the streams is closed. Is it assumed that this is implemented after removing the auto-release feature?

yutakahirano commented:
FYI: https://github.com/yutakahirano/fetch-with-streams/pull/48#issuecomment-125172550

domenic commented on index.bs:
```diff
@@ -300,6 +300,11 @@ Instances of <code>ReadableStream</code> are created with the internal slots des
   </tr>
   <tr>
+    <td>\[[disturbed]]
+    <td>A boolean flag set to <b>true</b> when either <code>read()</code> or <code>cancel()</code> has been called on
```
better phrased as "when the stream has been read from or canceled"

tyoshino replied:
Done

domenic commented on index.bs:
```diff
@@ -971,6 +978,16 @@ an assert).
 
+<h4 id="is-readable-stream-disturbed" aoid="IsReadableStreamDisturbed">IsReadableStreamDisturbed ( stream )</h4>
+
+This abstract operation is meant to be called from other specifications that may wish to query whether or not a
```
"...readable stream has ever been read from or canceled"

tyoshino replied:
Done

domenic commented on reference-implementation/run-tests.js:
```diff
@@ -2,12 +2,14 @@ const glob = require('glob');
 import TransformStream from './lib/transform-stream';
 
 global.ReadableStream = ReadableStream;
+global.IsReadableStreamDisturbed = IsReadableStreamDisturbed;
```
I don't like this testing strategy because the global modifications are meant to be based on https://streams.spec.whatwg.org/#globals. I am not sure what the best strategy for testing it is though. I will try to work on something.

domenic replied:
For example, such tests could never be run in browsers. Maybe they just should be left for other specs' test suites.

tyoshino replied:
So, should we remove this from run-tests.js and ... import in the test file?

domenic replied:
Yes, that sounds good. And maybe move them to a separate abstract-ops.js or from-other-specs.js test file.

tyoshino replied:
Done

domenic commented:
> The disturbed flag is never changed when the streams is closed. Is it assumed that this is implemented after removing the auto-release feature?

I don't think we need to remove auto-release. However, as @yutakahirano notes, I think it was important (for reasons I forget) to have the disturbed flag change even on closed streams. As such it should probably move up a few steps in the algorithms.

tyoshino commented:
Maybe one of the origins of the requirement is https://github.com/yutakahirano/fetch-with-streams/issues/37#issuecomment-108116997 ? Is it still active? https://github.com/whatwg/fetch/issues/61#issuecomment-123854971 Or even if it's active, it'll be covered by used flag?

domenic commented:
Reviewing https://etherpad.mozilla.org/streams-f2f-july "Even more conservative strawman" I think the requirement comes from wanting .json() etc. to automatically make the stream disturbed, even for empty streams. The used flag is only for `new Request(otherRequest)`. This is important because it means that author code that acts similar to json() will also make the stream disturbed.

tyoshino commented:
Addressed the comment. Thanks.

> This is important because it means that author code that acts similar to json() will also make the stream disturbed.

OK. But an operation made on a released reader making the ex-associated stream disturbed sounds strange.

domenic commented:
> But an operation made on a released reader making the ex-associated stream disturbed sounds strange.

Oh, you are right. It would mean that getting a reader for a closed/error stream would still have to hold on to the stream in order to maybe mark it disturbed.

This seems a bit silly. It means you cannot GC the stream if the reader is still alive, even though the stream is closed or errored.

But that is maybe not so bad since closed or errored streams do not have any data in their internal queue.

Also it would eliminate some internal complexity from the reader (the [[state]] and [[storedError]] fields).

And the silliness seems less bad than debating bodyUsed all over again.

What do you think?

domenic commented:
Ping. What are your thoughts?

yutakahirano commented:
I prefer dropping auto-release to introducing released-but-still-connected state.

domenic commented:
That seems like a good compromise to me. @tyoshino do you want to do the PR or shall I?

domenic commented:
Actually, I can do that; there is other work I want to do on top of it. I will send a PR later today.

domenic commented:
OK, I have pushed to this PR with a base commit that removes automatic unlocking on close/error.

Two important questions:

- Should we now reject read() and closed when used on released readers, instead of treating as a closed/errored stream? That seems simpler.

- **Are we OK with breaking compat on this?** It would be a breaking change to Chrome's current behavior, and a change in an area of the spec we've called stable. I personally think it is unlikely to break things, but what do you think? /cc @yutakahirano @calvaris

calvaris commented:
I think it is also unlikely that we break things.

Our current implementation is also based on closed/errored so it would mean to change our current implementation, which I'm not a big fan of but it wouldn't be a big change either, so I'm against but only barely and based on laziness :)

What I would do is securing the change of behavior if it happens with some tests ensuring that everything is as it should be and not as it was.

domenic commented:
@calvaris thanks for weighing in! I made a bunch of test updates in https://github.com/whatwg/streams/commit/9945ae13eea9c1156d1feae3bfe665d9f7a88fc4 but if you can think of others we should add them. E.g. if you come up with test descriptions I'll write the tests :)

yutakahirano commented:
I believe we cannot introduce IsDisturbed with satisfying all proposed requirements and all existing properties. I think dropping the auto-release is a tolerable way.

domenic commented:
@yutakahirano well, we can introduce it without any backward-incompatible changes if we follow https://github.com/whatwg/streams/pull/385#issuecomment-128149814 and allow readers to retain references to their owner stream even after release.

But I take it you think that is bad?

yutakahirano commented:
> But I take it you think that is bad?

Yes.

domenic commented:
OK. Does anyone have any opinion on

> Should we now reject read() and closed when used on released readers, instead of treating as a closed/errored stream?

domenic commented:
My opinion is that they should now reject. That will allow us to get rid of the internal slots [[state]] and [[storedError]]. The downside is that you would no longer get to treat released readers as closed/errored streams depending on their originating stream. Instead they would effectively behave as if they always came from errored streams. But I don't think that's a big deal; you shouldn't be messing with the reader after it's been released.

tyoshino commented:
Sorry for delay. I'm fine with the plan. I don't come up with any reasonable use case to keep a reference to a reader while releasing the parent stream. Some code may be accidentally doing so, but could be easily fixed and as Domenic said it's not so harmful to hold an errored/closed stream (which is small) in addition to a reader.

I'm also fine with making read on released reader reject. I think currently people who are unaware of the auto release mechanism are not calling release(), and basically people who are aware of wouldn't read() on released one. Some may be using release() and also using released reader as a representation of closed stream that accepts read() and tell us that the stream is closed repeatedly. But ... maybe the number of users with such a use case is small, I guess.

domenic commented:
OK. Updated with what I think should be final. Cannot build the spec until https://github.com/tabatkins/bikeshed/issues/466 is fixed, but ready for review.

tyoshino commented on index.bs:
```diff
@@ -855,7 +848,10 @@ Instances of <code>ReadableStreamReader</code> are created with the internal slo
   1. If *this*@[[readRequests]] is not empty, throw a *TypeError* exception.
-  1. Perform CloseReadableStreamReader(*this*).
+  1. If *this*@[[ownerReadableStream]]@[[state]] is `"readable"`, reject *this*@[[closedPromise]] with a *TypeError* exception.
+  1. Otherwise, let *this*@[[closedPromise]] be a new promise rejected with a *TypeError* exception.
```
I thought "let" is used for declaring/substituting to local variables while set is used for substituting new value to a slot. I'm just misunderstanding your style rule or this "let" should be "set"?

tyoshino replied:
Do we really need to set a new rejected promise to [[closedPromise]] here?

tyoshino replied:
Ah, I see. This new rejected promise and the TypeError at L851 are representing "already released". Then, looks good.

tyoshino commented on index.bs:
```diff
@@ -973,8 +958,6 @@ an assert).
     1. Reject _readRequestPromise_ with _e_.
   1. Set _reader_@[[readRequests]] to a new empty List.
   1. Perform ReleaseReadableStreamReader(_reader_).
```
Was the plan to keep auto-release for error?

tyoshino replied:
Oh, the impl is fixed. So, please just delete this line in the spec text :)

tyoshino commented on reference-implementation/lib/readable-stream.js:
```diff
@@ -562,13 +559,6 @@ function ReadFromReadableStreamReader(reader) {
   }
 }
 
-function ReleaseReadableStreamReader(reader) {
```
Let's remove ReleaseReadableStreamReader from the spec text too.

tyoshino commented:
Review done!

domenic commented:
Thanks for the review! I turns out I failed to do spec changes for the "Remove automatic unlocking on close/error" commit. Your let vs. set catch was correct as well. Fixed and will merge now.

yutakahirano commented:
Just a confirmation: @wanderview, are you OK with this change? I am planning to implement this behavior change to Chrome, and before that I would like to make sure that everyone is satisfied.

yutakahirano commented:
@wanderview: ping

wanderview commented:
I'm sorry, but I haven't been able to keep up with this.  Can you summarize
the differences from the July f2f agreement?

Are we just dropping automatic unlocking on stream close?
On Sep 10, 2015 10:18 PM, "Yutaka Hirano" <notifications@github.com> wrote:

> @wanderview <https://github.com/wanderview>: ping
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/whatwg/streams/pull/385#issuecomment-139431213>.
>

yutakahirano commented:
The f2f agreement suggests a stream should be disturbed  when read even if it is closed (See https://github.com/yutakahirano/fetch-with-streams/pull/48#issuecomment-125172550). In order to achieve that, there are two ways:

1. Auto-released reader makes the stream disturbed.
1. Remove the auto-release mechanism.

My opinion is that the latter is less evil.

> Are we just dropping automatic unlocking on stream close?

And on stream error.

wanderview commented:
> Remove the auto-release mechanism.

I think this is reasonable.  Thanks for summarizing for me.
